streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -854,16 +870,216 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
                                                                                  final Serde<V> valueSerde) {
-        return Stores.windowStoreBuilder(
-            Stores.persistentWindowStore(
-                joinName + "-store",
-                Duration.ofMillis(windows.size() + windows.gracePeriodMs()),
-                Duration.ofMillis(windows.size()),
-                true
-            ),
-            keySerde,
-            valueSerde
+        final WindowBytesStoreSupplier supplier = Stores.persistentWindowWithTimestampStore(
+            joinName + "-store",
+            Duration.ofMillis(windows.size() + windows.gracePeriodMs()),
+            Duration.ofMillis(windows.size()),
+            true
         );
+        return new WindowStoreBuilder<K, V>(supplier, null, null, Time.SYSTEM) {
+            final StoreBuilder<WindowStore<K, ValueAndTimestamp<V>>> inner = Stores.windowWithTimestampStoreBuilder(
+                supplier,
+                keySerde,
+                valueSerde
+            );
+
+            @Override
+            public WindowStoreBuilder<K, V> withCachingEnabled() {
+                inner.withCachingEnabled();
+                return this;
+            }
+
+            @Override
+            public WindowStoreBuilder<K, V> withCachingDisabled() {
+                inner.withCachingDisabled();
+                return this;
+            }
+
+            @Override
+            public WindowStoreBuilder<K, V> withLoggingEnabled(final Map<String, String> config) {
+                inner.withLoggingEnabled(config);
+                return this;
+            }
+
+            @Override
+            public WindowStoreBuilder<K, V> withLoggingDisabled() {
+                inner.withLoggingDisabled();
+                return this;
+            }
+
+            @Override
+            public Map<String, String> logConfig() {
+                return inner.logConfig();
+            }
+
+            @Override
+            public boolean loggingEnabled() {
+                return inner.loggingEnabled();
+            }
+
+            @Override
+            public String name() {
+                return inner.name();
+            }
+
+            @Override
+            public WindowStore<K, V> build() {
+                return new WindowStoreFacade<>(inner.build());
+            }
+        };
+    }
+
+    public static class WindowStoreFacade<A, B> implements WindowStore<A, B>, RecordConverter {
+        public final WindowStore<A, ValueAndTimestamp<B>> inner;
+        private final RecordConverter innerRecordConvert;
+
+        public WindowStoreFacade(final WindowStore<A, ValueAndTimestamp<B>> store) {
+            this.inner = store;
+            final StateStore rootStore = store instanceof WrappedStateStore ? ((WrappedStateStore) store).inner() : store;
+            innerRecordConvert = rootStore instanceof RecordConverter ? (RecordConverter) rootStore : new DefaultRecordConverter();
```

Comment:
This makes me thinking about `WrappedStateStore` in which we do not check inner but blindly assumes it always implement this interface:

```
        public ConsumerRecord<byte[], byte[]> convert(final ConsumerRecord<byte[], byte[]> record) {
            return ((RecordConverter) innerState).convert(record);
        }
```

Is that safe? Could it be a non-caching, non-logging store (i.e. only one layer of wrapper on metered, and the inner is already the inner-most underlying store which is customized store not implementing the interface)?

Reply:
That could be the case -- this part of the code is not extensively tested yet. I was aware of it already on the old "big PR" and planned to address it, on the smaller PR when we actually add upgrade tests ect.